unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp

```diff
@@ -0,0 +1,261 @@
+{
+  const typet char_type = unsignedbv_typet(16);
+  const typet length_type = signedbv_typet(32);
+  const std::size_t pointer_width = 16;
+  const auto pointer_type = pointer_typet(char_type, pointer_width);
+
+  const symbol_tablet symbol_table;
+  const namespacet ns{symbol_table};
+  symbol_generatort fresh_symbol;
+  array_poolt array_pool{fresh_symbol};
+
+  WHEN("format specifier = %s")
+  {
+    WHEN("arg = serialized(\"bar\")")
+    {
+      array_string_exprt arg = from_char_vector(
+        {'b', 'a', 'r'}, char_type, length_type, pointer_type, array_pool);
+      format_specifiert fs{0, "", 0, 0, false, format_specifiert::STRING};
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      THEN("length constraint represents 3")
+      {
+        int to_int = *numeric_cast<int>(to_constant_expr(actual));
+        REQUIRE(to_int == 3);
+      }
+    }
+  }
+
+  WHEN("format specifier = %d")
+  {
+    WHEN("arg = serialized(12)")
+    {
+      const array_string_exprt arg = from_char_vector(
+        {0, 0, 0, 12}, char_type, length_type, pointer_type, array_pool);
+      format_specifiert fs{
+        0, "", 0, 0, false, format_specifiert::DECIMAL_INTEGER};
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      const exprt simpl_actual = simplify_expr(actual, ns);
+      THEN("length constraint represents 2")
+      {
+        const int to_int = *numeric_cast<int>(to_constant_expr(simpl_actual));
+        REQUIRE(to_int == 2);
+      }
+    }
+  }
+
+  WHEN("format specifier = %d")
+  {
+    WHEN("arg = serialized(-12)")
+    {
+      const array_string_exprt arg = from_char_vector(
+        {0xFFFF, 0xFFFF, 0xFFFF, 0xFFF4},
+        char_type,
+        length_type,
+        pointer_type,
+        array_pool);
+      format_specifiert fs{
+        0, "", 0, 0, false, format_specifiert::DECIMAL_INTEGER};
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      const exprt simpl_actual = simplify_expr(actual, ns);
+      THEN("length constraint represents 2")
+      {
+        const int to_int = *numeric_cast<int>(to_constant_expr(simpl_actual));
+        REQUIRE(to_int == 3);
+      }
+    }
+  }
+
+  WHEN("format specifier = %d")
+  {
+    WHEN("arg = serialized(-1234567890)")
+    {
+      const array_string_exprt arg = from_char_vector(
+        {0xFFFF, 0xFFFF, 0xB669, 0xFD2E},
+        char_type,
+        length_type,
+        pointer_type,
+        array_pool);
+      format_specifiert fs{
+        0, "", 0, 0, false, format_specifiert::DECIMAL_INTEGER};
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      const exprt simpl_actual = simplify_expr(actual, ns);
+      THEN("length constraint represents 11")
+      {
+        const int to_int = *numeric_cast<int>(to_constant_expr(simpl_actual));
+        REQUIRE(to_int == 11);
```

Comment:
:pick: I would add a first repquire which check that the `numeric_cast` had a non-empty return

Reply:
I don't think that's very likely to happen. I'll keep it like this to keep the unit test simple. It will fail anyway if the conversion fails (although not as gracefully).